npm/networkPolicyController_test.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2018 Microsoft. All rights reserved.
+// MIT License
+package npm
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/Azure/azure-container-networking/npm/ipsm"
+
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	kubeinformers "k8s.io/client-go/informers"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	core "k8s.io/client-go/testing"
+)
+
+type netPolFixture struct {
+	t *testing.T
+
+	kubeclient *k8sfake.Clientset
+	// Objects to put in the store.
+	netPolLister []*networkingv1.NetworkPolicy
+	// (TODO) Actions expected to happen on the client. Will use this to check action.
+	kubeactions []core.Action
+	// Objects from here preloaded into NewSimpleFake.
+	kubeobjects []runtime.Object
+
+	// (TODO) will remove npMgr if possible
+	npMgr            *NetworkPolicyManager
+	ipsMgr           *ipsm.IpsetManager
+	netPolController *networkPolicyController
+	kubeInformer     kubeinformers.SharedInformerFactory
+}
+
+func newNetPolFixture(t *testing.T) *netPolFixture {
+	f := &netPolFixture{
+		t:            t,
+		netPolLister: []*networkingv1.NetworkPolicy{},
+		kubeobjects:  []runtime.Object{},
+		npMgr:        newNPMgr(t),
+		ipsMgr:       ipsm.NewIpsetManager(),
+	}
+
+	f.npMgr.RawNpMap = make(map[string]*networkingv1.NetworkPolicy)
+	return f
+}
+
+func (f *netPolFixture) newNetPolController(stopCh chan struct{}) {
+	f.kubeclient = k8sfake.NewSimpleClientset(f.kubeobjects...)
+	f.kubeInformer = kubeinformers.NewSharedInformerFactory(f.kubeclient, noResyncPeriodFunc())
+
+	f.netPolController = NewNetworkPolicyController(f.kubeInformer.Networking().V1().NetworkPolicies(), f.kubeclient, f.npMgr)
+	f.netPolController.netPolListerSynced = alwaysReady
+
+	for _, netPol := range f.netPolLister {
+		f.kubeInformer.Networking().V1().NetworkPolicies().Informer().GetIndexer().Add(netPol)
+	}
+
+	f.kubeInformer.Start(stopCh)
+}
+
+func (f *netPolFixture) ipSetSave(ipsetConfigFile string) {
+	//  call /sbin/ipset save -file /var/log/ipset-test.conf
+	f.t.Logf("Start storing ipset to %s", ipsetConfigFile)
+	if err := f.ipsMgr.Save(ipsetConfigFile); err != nil {
+		f.t.Errorf("ipSetSave failed @ ipsMgr.Save")
+	}
+}
+func (f *netPolFixture) ipSetRestore(ipsetConfigFile string) {
```

Comment:
We can reuse ipsetsave and restore from other tests right ?

Reply:
Definitely. we can. I would suggest we left them separately now and see them after reviewing and thinking overall structure of our codes since as we discussed, we may decouple each controller into each package later.